File: modules/yandex-map/android/src/main/java/expo/modules/yandexmap/view/PolylineView.kt
package expo.modules.yandexmap.view

import android.annotation.SuppressLint
import android.content.Context
import com.yandex.mapkit.geometry.Polyline
import com.yandex.mapkit.map.PolylineMapObject
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.views.ExpoView
import expo.modules.yandexmap.model.Coordinate
import expo.modules.yandexmap.model.PolylineConfig
import expo.modules.yandexmap.utils.ColorUtils
import expo.modules.yandexmap.view.YandexMapView.Companion.mapObjects

@SuppressLint("ViewConstructor")
class PolylineView(context: Context, appContext: AppContext) : ExpoView(context, appContext) {
  var polylineMapObject: PolylineMapObject? = null

  private var polylineConfig: PolylineConfig = PolylineConfig()

  fun setPointsValue(coordinates: List<Coordinate>) {
    polylineConfig.points = coordinates.map { point ->
      point.toPoint()
    }
  }

  fun setStrokeWidthValue(width: Double?) {
    if (width != null) {
      polylineConfig.strokeWidth = width.toFloat()
    }
  }

  fun setStrokeColorValue(color: String?) {
    if (color != null) {
      polylineConfig.strokeColor = ColorUtils.parseColor(color)
    }
  }

  fun setOutlineWidthValue(width: Double?) {
    if (width != null) {
      polylineConfig.outlineWidth = width.toFloat()
    }
  }

  fun setOutlineColorValue(color: String?) {
    if (color != null) {
      polylineConfig.outlineColor = ColorUtils.parseColor(color)
    }
  }

  fun updatePolyline() {
    if (polylineConfig.points != null) {
      val polyline = Polyline(polylineConfig.points!!)

      polylineMapObject = mapObjects?.addPolyline(polyline)?.apply {
        if (polylineConfig.strokeWidth != null) {
          strokeWidth = polylineConfig.strokeWidth!!
        }

        if (polylineConfig.strokeColor != null) {
          setStrokeColor(polylineConfig.strokeColor!!)
        }

        if (polylineConfig.outlineWidth != null) {
          outlineWidth = polylineConfig.outlineWidth!!
        }

        if (polylineConfig.outlineColor != null) {
          outlineColor = polylineConfig.outlineColor!!
        }
      }
    }
  }
}

File: modules/yandex-map/ios/View/PolylineView.swift
import ExpoModulesCore
import YandexMapsMobile

class PolylineView: ExpoView {
    weak var delegate: MapObjectsDelegate?
    
    var polylineMapObject: YMKPolylineMapObject? = nil
    
    private var polylineConfig: PolylineConfig = PolylineConfig()

    func setPointsValue(_ points: Array<Coordinate>) {
        let pointMap = points.map { $0.toPoint() }
        
        polylineConfig.points = pointMap
    }
    
    func setStrokeWidthValue(_ strokeWidth: Float) {
        polylineConfig.strokeWidth = strokeWidth
    }

    func setStrokeColorValue(_ color: String) {
        polylineConfig.strokeColor = color
    }
    
    func setOutlineWidthValue(_ strokeWidth: Float) {
        polylineConfig.outlineWidth = strokeWidth
    }

    func setOutlineColorValue(_ color: String) {
        polylineConfig.outlineColor = color
    }
    
    func updatePolyline() {
        guard let mapObjects = delegate?.mapObjects else { return }
        
        let polyline: YMKPolyline = {
            YMKPolyline(
                points: polylineConfig.points
            )
        }()
        
        polylineMapObject = mapObjects.addPolyline(with: polyline)
        
        if polylineConfig.strokeColor != nil {
            polylineMapObject!.setStrokeColorWith(UIColor.fromString(polylineConfig.strokeColor!)!)
        }
        
        if polylineConfig.outlineColor != nil {
            polylineMapObject!.outlineColor = UIColor.fromString(polylineConfig.outlineColor!)!
        }
        
        if polylineConfig.strokeWidth != nil {
            polylineMapObject!.strokeWidth = polylineConfig.strokeWidth!
        }
        
        if polylineConfig.outlineWidth != nil {
            polylineMapObject!.outlineWidth = polylineConfig.outlineWidth!
        }
    }
}

